File: src/leetcode/242.ValidAnagram.js
/*
Given two strings s and t, return true if t is an anagram of s, and false otherwise.
An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, 
typically using all the original letters exactly once.
Example:
Input: s = "rat", t = "tar"
Output: false 
*/

// O(n) T O(n) Space
const isAnagram = (s, t) => {
  if (s.length != t.length) return false;
  let charObj = {};
  for (char of s) {
    if (charObj[char]) {
      charObj[char]++;
    } else {
      charObj[char] = 1;
    }
  }
  for (char of t) {
    if (charObj[char]) {
      charObj[char]--;
    } else {
      return false;
    }
  }
  return true;
};

//tests
console.log(validAnagram1("anagram", "nagaram"));
console.log(validAnagram1("anagrm", "nagaram"));
console.log(validAnagram1("", "nagaram"));
console.log(validAnagram1("ab", "a"));

File: src/leetcode/python/242.ValidAnagram.py
'''
Given two strings s and t, return true if t is an anagram of s, and false otherwise.

An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase, typically using all the original letters exactly once.

 '''
class Solution:
    #Optimal sln using hashMap Time Complexity O(n) Space complexity O(n)
    def isAnagram(self, s: str, t:str) -> bool:
        #compare the length of the strings
        if len(s) != len(t):
            return False
        #hashMap to store the characters in the string
        sMap = {}
        tMap = {}

        #loop through the s string and check if the values are in the sMap
        for char in s:
            if char in sMap:
                sMap[char] +=1
            else:
                sMap[char] = 1
        
        #loop through the t string and check if the values are in the tMap
        for char in t:
            if char in tMap:
                tMap[char] +=1
            else:
                tMap[char] = 1
        
        #loop through the sMap
        for char in sMap:
            #return False is the values in sMap are not in tMap
            if char not in tMap or sMap[char] != tMap[char]:
                return False
        
        #return True if no return yet because they must be anagrams
        return True

s = Solution()
print(s.isAnagram("anagram", "nagaram"))    #True
print(s.isAnagram("rat", "car"))    #False
